prov/util/src/util_av.c

```diff
@@ -905,79 +885,99 @@ static int ip_av_insertsvc(struct fid_av *av, const char *node,
 	return fi_av_insertsym(av, node, 1, service, 1, fi_addr, flags, context);
 }
 
-static int ip_av_insert_ip4sym(struct util_av *av,
-			       struct in_addr ip, size_t ipcnt,
-			       uint16_t port, size_t portcnt,
-			       fi_addr_t *fi_addr, void *context)
+/* Caller should free *addr */
+static int
+ip_av_ip4sym_getaddr(struct util_av *av, struct in_addr ip, size_t ipcnt,
+		     uint16_t port, size_t portcnt, void **addr)
 {
-	struct sockaddr_in sin;
-	int fi, ret, success_cnt = 0;
-	size_t i, p;
+	struct sockaddr_in *sin;
+	int count = ipcnt * portcnt;
+	size_t i, p, k;
 
-	memset(&sin, 0, sizeof sin);
-	sin.sin_family = AF_INET;
-
-	for (i = 0, fi = 0; i < ipcnt; i++) {
-		/* TODO: should we skip addresses x.x.x.0 and x.x.x.255? */
-		sin.sin_addr.s_addr = htonl(ntohl(ip.s_addr) + i);
+	sin = calloc(count, sizeof(*sin));
+	if (!sin)
+		return -FI_ENOMEM;
 
-		for (p = 0; p < portcnt; p++, fi++) {
-			sin.sin_port = htons(port + p);
-			ret = ip_av_insert_addr(av, &sin, fi_addr ?
-						&fi_addr[fi] : NULL, context);
-			if (!ret)
-				success_cnt++;
-			else if (av->eq)
-				ofi_av_write_event(av, fi, -ret, context);
+	for (i = 0; i < ipcnt; i++) {
+		for (p = 0; p < portcnt; p++) {
+			k = i * portcnt + p;
+			sin[k].sin_family = AF_INET;
+			/* TODO: should we skip addresses x.x.x.0 and x.x.x.255? */
+			sin[k].sin_addr.s_addr = htonl(ntohl(ip.s_addr) + i);
+			sin[k].sin_port = htons(port + p);
```

Comment:
and increment k++ here?

This series of comments applies to the ipv6 version of this function